File: user-impl/src/main/java/com/gabizou/cameraiq/demo/impl/UserEntity.java
package com.gabizou.cameraiq.demo.impl;

import com.gabizou.cameraiq.demo.api.UserId;
import com.gabizou.cameraiq.demo.impl.events.UserEvent;
import com.lightbend.lagom.javadsl.persistence.PersistentEntity;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.pcollections.OrderedPSet;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Optional;


public final class UserEntity extends PersistentEntity<UserCommand, UserEvent, UserState> {

    private static final Logger LOGGER = LogManager.getLogger(UserEntity.class);

    @Override
    public Behavior initialBehavior(final Optional<UserState> snapshotState) {
        final BehaviorBuilder behavior =
            this.newBehaviorBuilder(snapshotState.orElseGet(() -> new UserState(OrderedPSet.empty())));
        behavior.setCommandHandler(UserCommand.CreateUser.class, (cmd, ctx) -> {
            final UserId userId = cmd.registration.userId;
            if (this.state().users.contains(userId)) {
                UserEntity.LOGGER.warn("No user found by id: " + userId);
                ctx.invalidCommand("User " + this.entityId() + " is already created");
                return ctx.done();
            }
            final ArrayList<UserEvent> events = new ArrayList<>();
            // Actually add the user based on registration

            events.add(new UserEvent.UserCreated(userId, Instant.now()));

            return ctx.thenPersistAll(events, () -> ctx.reply(cmd.registration));
        });
        behavior.setEventHandler(UserEvent.UserCreated.class,
            event -> {
                UserEntity.LOGGER.debug("Updating User State with (" + this.state().users.size() + 1 + ") users.");
            return new UserState(this.state().users.plus(event.userId));
            });
        behavior.setReadOnlyCommandHandler(UserCommand.GetUser.class, (cmd, ctx) -> {
            final UserId userId = cmd.uuid;
            if (!this.state().users.contains(userId)) {
                ctx.invalidCommand("User " + cmd.uuid + " does not " + "exist");
            }
            UserEntity.LOGGER.debug("Fetching registered user " + userId);

        });
        return behavior.build();
    }
}

File: user-impl/src/main/scala/com/gabizou/demo/impl/UserEntity.scala
package com.gabizou.demo.impl

import java.time.Instant

import akka.Done
import com.gabizou.demo.{UserCreated, UserEvent}
import com.gabizou.demo.api.{User, UserId}
import com.lightbend.lagom.scaladsl.persistence.PersistentEntity
import org.apache.log4j.LogManager

class UserEntity extends PersistentEntity {
  override type Command = UserCommand[_]
  override type Event = UserEvent
  override type State = UserState

  sealed trait UserState {
    def users: Set[UserId]
    def containsUser(userId: UserId): Boolean
  }
  case object EmptyState extends UserState {
    override def containsUser(userId: UserId): Boolean = false

    override def users: Set[UserId] = Set.empty
  }
  case class PopulatedState(set: Set[UserId]) extends UserState {
    override def containsUser(userId: UserId): Boolean = set.contains(userId)

    override def users: Set[UserId] = set
  }
  override def initialState: UserState = EmptyState

  override def behavior: Behavior = {
    case EmptyState => Actions().onCommand[CreateUser, User] {
      case (command: CreateUser, value, state) =>
        val id = command.registration.userId
        if (state.containsUser(id)) {
          UserEntity.LOGGER.warn("Duplicate user found by id: " + id)
          value.invalidCommand(s"User $id is already created")
          value.done
        }
        value.thenPersist(UserCreated(id, Instant.now)) { _ =>
          value.reply(command.registration)
        }
    }
        .onEvent {
          case (UserCreated(userId, _), state) =>
            UserEntity.LOGGER.debug("Updating User State with (" + state.users.size + 1 + ") users.")
            PopulatedState(state.users + userId)
        }
      .onReadOnlyCommand[GetUser, Done] {
      case(command:GetUser, value, state) =>
        val id = command.userId
        if (!state.containsUser(id)) {
          value.invalidCommand(s"User $id does not exist")
        }
        UserEntity.LOGGER.debug(s"Fetching registered user $id")
        value.reply(Done)
      }
  }
}
object UserEntity {
  private val LOGGER = LogManager.getLogger(classOf[UserEntity])

}

case class UserState(set: Set[UserId]) {

}